### PR TITLE
Determine prunability of tables for join operations

### PR DESCRIPTION
**Pull request by berkaysynnada**
# Which issue does this PR close?

Closes #.

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/intervals/prunability.rs [hunk 1]
```diff
@@ -0,0 +1,802 @@
...
+                    .clone()
+                    .as_any()
+                    .downcast_ref::<Column>()
+                {
+                    self.set_leaf_column(column, sort_exprs, node, left_schema);
+                }
+            }
+            // intermediate nodes
+            else {
+                children_order.reverse();
```
Comment (mustafasrepo) [inline datafusion/physical-expr/src/intervals/prunability.rs:132]: Why do we need to reverse add comment

File: datafusion/physical-expr/src/intervals/prunability.rs [hunk 2]
```diff
@@ -0,0 +1,802 @@
...
+                    opt.0,
+                    SortOptions {
+                        descending: !opt.1.descending,
+                        nulls_first: opt.1.nulls_first,
+                    },
+                )
+            })
+        }
+        // Some column + - literal
+        (_, Some(_), _) => *children_order[0],
```
Comment (mustafasrepo) [inline datafusion/physical-expr/src/intervals/prunability.rs:297]: How can we sure that operation will only be minus or plus.

File: datafusion/physical-expr/src/intervals/prunability.rs [hunk 3]
```diff
@@ -0,0 +1,802 @@
...
+        binary_expr.right().as_any().downcast_ref::<Literal>(),
+        binary_expr.op(),
+    ) {
+        // Literal cmp some column or Some column cmp literal
+        (Some(_), _, _) | (_, Some(_), _) => None,
+        // Column cmp column
+        (_, _, op) => match (children_order[0], children_order[1], op) {
+            // Ordered + ordered column
+            (Some(left), Some(right), op) => {
+                match (left.1.descending, right.1.descending, op) {
```
Comment (mustafasrepo) [inline datafusion/physical-expr/src/intervals/prunability.rs:384]: Did you consider information in the nulls_first field? What happens in these cases?

**Comment by berkaysynnada** [comment]
change input sort expr as left_sort_expr and right_sort_expr

File: datafusion/physical-expr/src/intervals/prunability.rs [hunk 4]
```diff
@@ -0,0 +1,798 @@
...
+        Ok(Self { graph, root })
+    }
+
+    /// This function determines the prunability of the given
+    /// ExprPrunabilityGraph, with PhysicalSortExpr's of some columns.
+    /// The result is returned with a tuple, the first one is the prunability
+    /// of the first table, and the second one is of the second table.
+    pub fn is_prunable(
+        &mut self,
+        sort_exprs: &[&PhysicalSortExpr],
```
Comment (berkaysynnada) [inline datafusion/physical-expr/src/intervals/prunability.rs:104]: separate this as left and right

File: datafusion/physical-expr/src/intervals/prunability.rs [hunk 5]
```diff
@@ -0,0 +1,950 @@
...
+            graph.analyze_prunability(
+                &Some(left_sorted1_desc),
+                &Some(right_sorted1_asc),
+            )?
+        );
+
+        Ok(())
+    }
+
+    fn experiment_sort_info(
```
Comment (mustafasrepo) [inline datafusion/physical-expr/src/intervals/prunability.rs:730]: Add comment for this function, such as "expect its input expression `expr` to be in the form: "

File: datafusion/physical-expr/src/intervals/prunability.rs [hunk 6]
```diff
@@ -0,0 +1,950 @@
...
+                    nulls_last: false
+                }
+            ),
+            graph.analyze_prunability(&Some(left_sorted_asc), &None,)?
+        );
+
+        Ok(())
+    }
+
+    fn experiment_prunability_four_columns(
```
Comment (mustafasrepo) [inline datafusion/physical-expr/src/intervals/prunability.rs:618]: Add comment for this function, such as "expect its input expression `expr` to be in the form: "

File: datafusion/physical-expr/src/intervals/prunability.rs [hunk 7]
```diff
@@ -0,0 +1,950 @@
...
+#[cfg(test)]
+mod tests {
+    use std::ops::Not;
+
+    use super::*;
+    use crate::expressions::{col, BinaryExpr, Literal};
+    use arrow_schema::{DataType, Field, Schema, SortOptions};
+    use datafusion_common::ScalarValue;
+
+    fn experiment_prunability(
```
Comment (mustafasrepo) [inline datafusion/physical-expr/src/intervals/prunability.rs:525]: Add comment for this function, such as "expect its input expression `expr` to be in the form: "

**Review by mustafasrepo — APPROVED** [review]

**Review by ozankabak — APPROVED** [review]
LGTM, sent in a commit with some simplifications and readability changes. Remaining steps I can think of before we can send upstream:
1. Writing an explanatory PR body.
2. Hooking this up with the `is_prunable` check in `PipelineChecker`.

If there is anything I'm missing, let's take those steps as well.

File: datafusion/physical-expr/src/intervals/prunability.rs [hunk 8]
```diff
@@ -0,0 +1,892 @@
...
+                .collect::<Vec<_>>();
+
+            // Handle leaf nodes:
+            if children_prunability.is_empty() {
+                // Initialize as an unordered column with an undefined table side:
+                self.graph[node].table_side = TableSide::None;
+                self.graph[node].sort_info.dir = Monotonicity::Unordered;
+                // By convention, we initially set nulls_first and nulls_last to false.
+                self.graph[node].sort_info.nulls_first = true;
+                self.graph[node].sort_info.nulls_last = false;
```
Comment (ozankabak) [inline datafusion/physical-expr/src/intervals/prunability.rs:280]: The comment and the assignment are contradictory. Let's fix the wrong one.

**Pull request merged**